pylint/test/functional/consider_join.py

```diff
@@ -0,0 +1,110 @@
+# pylint: disable=missing-docstring,invalid-name,undefined-variable,multiple-statements
+
+# Variations of 'result'
+result = ''
+for number in ['1', '2', '3']:
+    result += number  # [consider-using-join]
+
+result = 'header'
+for number in ['1', '2', '3']:
+    result += number  # [consider-using-join]
+
+result = another_result = ''
+for number in ['1', '2', '3']:
+    result += number  # [consider-using-join]
+
+another_result = result = ''
+for number in ['1', '2', '3']:
+    result += number  # [consider-using-join]
+
+result = 0  # result is not a string
+for number in ['1', '2', '3']:
+    result += number
+
+RESULT = ''  # wrong name / initial variable missing
+for number in ['1', '2', '3']:
+    result += [number]
+
+string_variable = ''
+result = string_variable  # type of 'result' not obviously a string
+for number in ['1', '2', '3']:
+    result += number
+
+result = ''
+another_result = ''  # result defined too early
+for number in ['1', '2', '3']:
+    result += [number]
+
+for number in ['1', '2', '3']:  # 'result'-definition missing
+    result += number
+
+
+# Variations of 'number'
+result = ''  # no concatenation (iterator-name differs)
+for name in ['1', '2', '3']:
+    result += number
+
+result = ''  # no concatenation (iterator-name differs)
+for _ in ['1', '2', '3']:
+    result += number
+# 'exprlist' is not a single name
+for index, number in ['1', '2', '3']:
+    result += number
+
+
+# Variations of 'iterable'
+result = ''
+for number in []:
+    result += number  # [consider-using-join]
+
+result = ''
+for number in "a text":
+    result += number  # [consider-using-join]
+
+result = ''
+for number in [1, 2, 3]:
+    result += number  # [consider-using-join]
```

Comment:
''.join([1,2,3]) raises an error

Reply:
I think this is fine. We'd have to inspect the elements of the iterable which complicate the patch a little bit. @pheanex what we could do though is to look for the type of the assignment before the loop. If that is not a string, we shouldn't emit this check.

Reply:
@brycepg yes I considered to check the inspection of the variables in the iterable, but also thought that this might not be worth the effort, since most of the time the iterable is a variable, and checking this as well is even harder. Do we already have the capability to check a type of a variable during a lint-run? And how reliable would that even be?

@PCManticore Haven't I implemented this check already (See line 20-22) in the functional tests or line 556 (isinstance(assign.value.value, str)) in the checker?

Reply:
If you're referring to inferring the type of a variable, that's doable and works even today. Have you tried running an `.infer()` over the iterable if that's what you're asking? if you want the actual type you want try with `astroid.helpers.object_type`.
Ah cool! Nice that you defaulted to action.

Reply:
Oh ok cool, I wasn't aware of ".infer()". Let me give it a try to see if I can make the checker more generic with it. :smile: